### PR TITLE
learn2016: i18n

### PR DESCRIPTION
**Pull request by breville**
i18n support for all strings in tutorialExplorer app.

Note that there's no support for localising strings that are coming from the HocTutorials ghseet.

### CODE REVIEW

File: apps/Gruntfile.js [hunk 1]
```diff
@@ -591,7 +591,7 @@ module.exports = function (grunt) {
   grunt.registerTask('locales', function () {
     var current = path.resolve('build/locale/current');
     mkdirp.sync(current);
-    appsToBuild.concat('common').map(function (item) {
+    appsToBuild.concat('common').concat('tutorialExplorer').map(function (item) {
```
Comment (Bjvanminnen) [inline apps/Gruntfile.js]: This can just be `appsToBuild.concat('common', 'tutorialExplorer').map(`

**Comment by Bjvanminnen** [comment]
Did this start working for you?

File: apps/src/tutorialExplorer/locale.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+/**
+ * DO NOT IMPORT THIS DIRECTLY. Instead do:
+ *   ```
+ *   import i18n from '@cdo/tutorialExplorer/locale'.
+ *   ```
+ * This allows the webpack config to determine how locales should be loaded,
+ * which is important for making locale setup work seamlessly in tests.
+ */
+// locale for tutorialExplorer
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/locale.js:1]: Remove lines 1-8, as it doesnt apply to this file now.

File: apps/i18n/tutorialExplorer/en_us.json [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  "filterProgrammingLanguage": "Language",
+  "filterProgrammingLanguageBlocks": "Blocks",
+  "filterProgrammingLanguageTyping": "Typing",
+  "filterProgrammingLanguageOther": "Other",
+  "filterActivityTypeRobotics": "Robotics",
+  "backButtonBack": "Back to all tutorials",
+  "filterHeaderFilterBy": "Filter By",
+  "filterHeaderShowFilters": "Filters",
+  "filterHeaderHideFilters": "Apply",
+  "roboticsButtonText": "Got robots? <a href='/learn/robotics'>Use these activities</a> and make a tangible Hour of Code for students of any age!",
```
Comment (Bjvanminnen) [inline apps/i18n/tutorialExplorer/en_us.json]: Do we need to be careful about links inside of translatable strings? (I think the answer might be yes). Nothing prevents others from "translating" this into a string with a different link.

File: apps/src/tutorialExplorer/roboticsButton.jsx [hunk 4]
```diff
@@ -34,7 +35,7 @@ const RoboticsButton = React.createClass({
           </a>
         </div>
         <div style={roboticsTextStyle}>
-          Got robots? <a href="/learn/robotics">Use these activities</a> and make a tangible Hour of Code for students of any age!
+          {i18n.roboticsButtonText()}
```
Comment (Bjvanminnen) [inline apps/src/tutorialExplorer/roboticsButton.jsx]: I'd make extra sure this works correctly. I would have guess React doesn't let you do this without using `dangerouslySetInnerHtml`

File: pegasus/sites.v3/code.org/public/learn/index.haml [hunk 5]
```diff
@@ -42,8 +42,10 @@ social:
         %br/
       #studentimage.col-50.tablet-feature{style: "height: 278px"}
 
--# Not ideal but we can pull the undigested file from /blockly.
-%script{src: CDO.studio_url('/blockly/js/tutorialExplorer.js')}
+-# Not ideal but we can pull the undigested files from /blockly.
+- js_locale = request.locale.to_s.downcase.tr('-', '_')
```
Comment (Bjvanminnen) [inline pegasus/sites.v3/code.org/public/learn/index.haml:51]: I'd be a little worried about whether this is sufficient for all locales. Might be worth checking with Tanya.

File: apps/i18n/tutorialExplorer/en_us.json (hunk 3 above)
Comment (tanyaparker) [reply apps/i18n/tutorialExplorer/en_us.json]: Usually I put a param like `<a href='%{robotics_url}'>` and then you can do `I18n.t(:roboticsButtonText, robotics_url: '/learn/robotics')`.

File: apps/i18n/tutorialExplorer/en_us.json [hunk 6]
```diff
@@ -0,0 +1,46 @@
+{
+  "filterGrades": "Grades",
+  "filterGradesPre": "Pre-reader",
+  "filterGrades25": "Grades 2-5",
+  "filterGrades68": "Grades 6-8",
+  "filterGrades9": "Grades 9+",
+  "filterTeacherExperience": "Educator's experience",
```
Comment (tanyaparker) [inline apps/i18n/tutorialExplorer/en_us.json]: Educator experience

I listed these text changes as P3 for later, but may be easier to just update here already?

File: apps/i18n/tutorialExplorer/en_us.json [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  "filterTeacherExperienceBeginner": "Beginner",
+  "filterTeacherExperienceComfortable": "Comfortable",
+  "filterStudentExperience": "Student's experience",
```
Comment (tanyaparker) [inline apps/i18n/tutorialExplorer/en_us.json]: Student experience

File: apps/i18n/tutorialExplorer/en_us.json [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  "filterStudentExperienceBeginner": "Beginner",
+  "filterStudentExperienceComfortable": "Comfortable",
+  "filterPlatform": "Classroom technology",
+  "filterPlatformComputers": "Computers",
+  "filterPlatformAndroid": "Android",
+  "filterPlatformIos": "iPad/iPhone",
+  "filterPlatformNoInternet": "Poor or no internet",
+  "filterPlatformNoComputers": "No computers or devices",
+  "filterTopics": "Topics",
```
Comment (tanyaparker) [inline apps/i18n/tutorialExplorer/en_us.json:19]: Topic

File: apps/i18n/tutorialExplorer/en_us.json [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  "filterProgrammingLanguageOther": "Other",
+  "filterActivityTypeRobotics": "Robotics",
+  "backButtonBack": "Back to all tutorials",
+  "filterHeaderFilterBy": "Filter By",
+  "filterHeaderShowFilters": "Filters",
+  "filterHeaderHideFilters": "Apply",
+  "roboticsButtonText": "Got robots? <a href='/learn/robotics'>Use these activities</a> and make a tangible Hour of Code for students of any age!",
+  "tutorialDetailsTeacherNotes": "Teacher notes",
+  "tutorialDetailsShortLink": "Short link",
+  "tutorialDetailInternationalLanguages": "International languages"
```
Comment (tanyaparker) [inline apps/i18n/tutorialExplorer/en_us.json]: Languages

File: pegasus/sites.v3/code.org/public/learn/index.haml (hunk 5 above)
Comment (tanyaparker) [reply pegasus/sites.v3/code.org/public/learn/index.haml:51]: What do you mean? Like whether this is the proper way to get the locale for all locales?

Comment (Bjvanminnen) [reply pegasus/sites.v3/code.org/public/learn/index.haml:51]: Yeah. Looking at how js_locale works in dashboard, it seems like this should be good.

**Comment by tanyaparker** [comment]
We may need to chat about this more, but I think this only localizes for code.org and csedweek.org. Hourofcode.com follow a different pipeline, but it'd be a shame to have to double localization files. :P

**Comment by tanyaparker** [comment]
We need to create the other locale files even if they're exactly copies of the English file. I can do it, but it'll add ~60 files so let me know when you're done with this branch and if it's ok for me to commit to this branch.

**Comment by tanyaparker** [comment]
We need to create the other locale files even if they're exact copies of the English file. I think that's why I get `Error: Cannot read property 'tutorialExplorer_locale' of undefined` error when I test in another locale. I can do it, but it'll add ~60 files so let me know when you're done with this branch and if it's ok for me to commit to this branch.

File: pegasus/sites.v3/code.org/public/learn/index.haml (hunk 5 above)
Comment (tanyaparker) [reply pegasus/sites.v3/code.org/public/learn/index.haml:51]: Yep, I think this is right. :)

File: apps/i18n/tutorialExplorer/en_us.json (hunk 8 above)
Comment (breville) [reply apps/i18n/tutorialExplorer/en_us.json:19]: Singular?  Spec still lists Topics.

**Comment by breville** [comment]
@tanyaparker all three domains pull from the same tutorialExplorer_locale.js on code.org, so I don't think we should need a speparate loc path.

Though the HAML-based strings (say for banners) might, but I'll do them in a separate PR.

Would it be possible for you to commit the other 60 files in a separate PR first?  That way it'll keep this PR a bit more readable for future reference.

File: i18n/locales/source/blockly-mooc/tutorialExplorer.json [hunk 10]
```diff
@@ -0,0 +1,46 @@
+{
```
Comment (breville) [inline i18n/locales/source/blockly-mooc/tutorialExplorer.json]: @tanyaparker curious: why did you add this one file?

**Comment by tanyaparker** [comment]
haha oops I was testing the sync-apps script to see if it would create those other locales. It didn't. I was just testing es_es.json so it already existed and part of the script automatically pushes. Sorry! Let me remove those files and add all locales in separate PR.

File: pegasus/helpers/hourofcode_helpers.rb [hunk 11]
```diff
@@ -93,6 +93,33 @@ def hoc_detect_language
...
+    return nil unless locale = HOC_LOCALES.find{|i| i == language || i.split('-').first == language}
+    parts = locale.split('-')
+    return "#{parts[0].downcase}-#{parts[1].upcase}"
+  end
+end
+
+def hoc_locale
+  language_to_locale(@language)
+end
+
```
Comment (breville) [inline pegasus/helpers/hourofcode_helpers.rb:122]: The functions above here were adapted from https://github.com/code-dot-org/code-dot-org/blob/staging/shared/middleware/varnish_environment.rb#L34.  Curious if there's a better way to reuse that code.

Comment (breville) [reply pegasus/helpers/hourofcode_helpers.rb:122]: @tanyaparker Will this generate the correct locale list that correlates to what's going to appear in apps/i18n/tutorialExplorer?
